### PR TITLE
add ISO C++ Guideline Support Library (gsl-lite)

### PR DESCRIPTION
**Pull request by apwojcik**
Currently, only ONNXIFI is using it, but in the future, nGraph may benefit, too.

The Guideline Support Library (GSL) contains functions and types that are suggested for use by the _C++ Core Guidelines_ maintained by the _Standard C++ Foundation_. The library includes types like `owner<>`, `not_null<>`, `span<>`, `string_span` and much more.

The `gsl-lite` recognizes when it is compiled for the CUDA platform and decorates functions (methods) with `__host__` and `__device__`.

Refer to https://github.com/isocpp/CppCoreGuidelines for more details.

Signed-off-by: Artur Wojcik <artur.wojcik@intel.com>

### CODE REVIEW

**Comment by apwojcik** [comment]
For starter, I would like to use `gsl::span<>`, it is described here https://en.cppreference.com/w/cpp/container/span. Trained ONNX models may have a gigabyte or more in size. I want to view the memory as ONNX model in form of a set of objects without copying the data.

TL;DR:
The _Cpp Core Guidelines_ is a set of core guidelines for modern C++, the recommendations from the standard committee signed by Bjarne Stroustrup and Herb Sutter, taking likely future enhancements and ISO Technical Specifications (TSs) into account. The aim is to help C++ programmers to write simpler, more efficient, more maintainable code. The guidelines are not mandatory. We may adopt only those we think we benefit the most.

The _Guidelines Support Library_ is the implementation of the guidelines in form of class templates. `gsl-lite` is single header file implementation of the library based on Microsoft's GSL.

**Comment by apwojcik** [comment]
Here's the link to the header file:
https://github.com/martinmoene/gsl-lite/blob/master/include/gsl/gsl-lite.hpp

**Comment by apwojcik** [comment]
I asked Jim John to start the approval process.

**Comment by diyessi** [comment]
The Microsoft version wants C++14 support, and we have requirements for older compilers.

**Comment by diyessi** [comment]
What is the status of this (getting approval, etc.)?
